### ISSUE TITLE
3D convolution and pooling for R

### ISSUE THREAD

**Issue by EddieIT** [dmlc/mxnet#5578]
I would like to know if 3D convolution and pooling is available for R? The convolution can take in kernel and stride with 3 dimension but the pooling is not able to do so. Further more, when I tried to use the infer_shape() function to verify the shape of the convolution, the function only accept data with 4 dimension, which is 1 dimension less for a 3D convolution. 

Would appreciate a if anyone can show me how to enable 3D convolution and pooling in R.

**Comment by thirdwing** [comment]
@EddieIT Can you also show your code? I will look into it.

**Comment by haimeh** [comment]
@EddieIT @thirdwing 
#I think this script recreates the problem?

require(mxnet)
#===============Example Data=================
dataGen <- function(a,b,c,d,e){data <- as.logical(c(round(runif(9000)-a,0),
                                                    round(runif(9000)-b,0),
                                                    round(runif(9000)-c,0),
                                                    round(runif(9000)-d,0),
                                                    round(runif(9000)-e,0)))
return(data)}

class0 <-dataGen(.1,.1,.5,.1,-.2)
class1 <-dataGen(.1,.4,.5,-.4,.1)
class2 <-dataGen(.5,-.1,.2,-.1,.1)
train.array <- rbind(class0,class1,class2)
dim(train.array) <- c(3,3,5,3000)
train.y <- as.array(c(rep(0,1000),rep(1,1000),rep(2,1000)))


trainIter = mx.io.arrayiter(data = train.array, label = train.y)


# ===============Symbolic Model=================
data <- mx.symbol.Variable(name='data')
drop1 <- mx.symbol.Dropout(data=data,p=.1)
# first conv
conv1 <- mx.symbol.Convolution(data=drop1, num_filter=50, kernel=c(1,1,5))
relu1 <- mx.symbol.LeakyReLU(data=conv1)
# second conv
conv2 <- mx.symbol.Convolution(data=relu1, num_filter=30, kernel=c(2,2,1))
relu2 <- mx.symbol.LeakyReLU(data=conv2)
# flat
flatten <- mx.symbol.Flatten(data=relu2)
fc <- mx.symbol.FullyConnected(data=flatten, num_hidden=300)
drop2 <- mx.symbol.Dropout(data=fc,p=.5)
# final fullc
fcOut <- mx.symbol.FullyConnected(data=drop2, num_hidden=3)
myNet <- mx.symbol.SoftmaxOutput(fcOut)

# ===============Generate Solver================
model <- mx.model.FeedForward.create(symbol=myNet,
                                     array.layout='auto',
                                     X=trainIter,
                                     ctx=mx.cpu(),
                                     num.round=100, 
                                     wd=0.00001,
                                     initializer=mx.init.Xavier(),
                                     eval.metric=mx.metric.accuracy,
                                     learning.rate=0.01, 
                                     optimizer = 'adagrad',
                                     verbose = T)
# Resulting Error:
#   [11:27:22] D:\Program Files (x86)\Jenkins\workspace\mxnet\mxnet\dmlc-core\include\dmlc/logging.h:300: [11:27:22] d:\program files (x86)\jenkins\workspace\mxnet\mxnet\src\operator\./convolution-inl.h:423: Check failed: dshp.ndim() == 5 (4 vs. 5) Input data should be 5D in batch-num_filter-depth-y-x
# Error in symbol$infer.shape(list(...)) : 
#   Error in operator convolution22: [11:27:22] d:\program files (x86)\jenkins\workspace\mxnet\mxnet\src\operator\./convolution-inl.h:423: Check failed: dshp.ndim() == 5 (4 vs. 5) Input data should be 5D in batch-num_filter-depth-y-x

**Comment by thirdwing** [comment]
@haimeh Sorry for the late reply. You can get the information from the python API document: http://mxnet.io/api/python/symbol.html#mxnet.symbol.Convolution

In 3D convolution, the shape of input should be `(batch_size, channel, depth, height, width)`.

You can try to change `dim(train.array) <- c(3,3,5,3000)` to `dim(train.array) <- c(3,3,5,1,3000)`.
